Sources/UIExtensions/Components/SwiftUI/UIKit Wrappers/PageView.swift

```diff
@@ -0,0 +1,238 @@
+
+        // This is the source-of-truth for collection of items.
+        fileprivate var items: DataCollection
+
+        // Cached element that is being presented.
+        var lastKnownSelectedIndex: DataCollection.Index?
+
+        fileprivate lazy var pageDataSource: PageControllerDataSource = {
+            PageControllerDataSource(
+                pageBefore: { [weak self] viewController in self?.pageBefore(viewController: viewController) },
+                pageAfter: { [weak self] viewController in self?.pageAfter(viewController: viewController) },
+                didFinishAnimating: { [weak self] pageViewController, finished, _, completed in
+                    guard let self = self,
+                        finished,
+                        completed,
+                        let visibleViewController = pageViewController.viewControllers?.first,
+                        let hostingController = visibleViewController as? IdentifiableHostingController<DataCollection.Element.ID, PageType>,
+                        let index = self.items.firstIndex(where: { $0.id == hostingController.id }) else { return }
+
+                    self.lastKnownSelectedIndex = index
+                    self.parent.activePageIndex = index
+                }
+            )
+        }()
+
+        init(
+            parent: PageView,
+            items: DataCollection,
+            initialIndex: DataCollection.Index?
+        ) {
+            self.parent = parent
+            self.items = items
+            self.lastKnownSelectedIndex = initialIndex
+        }
+
+        func pageBefore(viewController: UIViewController) -> UIViewController? {
+            guard let hostingController = viewController as? IdentifiableHostingController<DataCollection.Element.ID, PageType> else {
+                return nil
+            }
+
+            guard let index = items.firstIndex(where: { $0.id == hostingController.id }) else { return nil }
+
+            let previousIndex = items.index(before: index)
+            guard previousIndex >= items.startIndex && previousIndex < items.endIndex else { return nil }
+            return self.viewController(for: items[safe: previousIndex])
+        }
+
+        func pageAfter(viewController: UIViewController) -> UIViewController? {
+            guard let hostingController = viewController as? IdentifiableHostingController<DataCollection.Element.ID, PageType> else {
+                return nil
+            }
+
+            guard let index = items.firstIndex(where: { $0.id == hostingController.id }) else { return nil }
+
+            let nextIndex = items.index(after: index)
+            guard nextIndex >= items.startIndex && nextIndex < items.endIndex else { return nil }
+            return self.viewController(for: items[safe: nextIndex])
+        }
+
+        func viewController(for element: DataCollection.Element?) -> UIViewController {
+            guard let element = element else { return UIViewController() }
```

Comment:
If, for some reason, you return this `UIViewController()` here at one point, I think this whole control will break as this is not castable to IdentifiableHostingController anymore, breaking all pageBefore and pageAfter checks. 

Would returning nil be better?